### DISCUSSION TITLE
Let's talk tree view layout

### DISCUSSION THREAD

**Discussion by ykud** [bgregs514/vscode-tm1#10] [General]
Was debating whether it's an issue or discussion, but it is a discussion indeed :) 

So I was thinking that the tree view we're exposing on the left pane should look like following with ^ showing collapsable tree nodes. It will make it quite useful from my point of view, but I'd want to hear whether it makes sense before I spend time implementing something along these lines.

* Server ^
  * Cubes ^
    * Cube
      * Rule
      * Views ^
        * View -- edit view definition
     * Edit structure?
  * Processes ^
     * Process ^
       * Run - maybe?
       * Edit
   * Dimensions ^
      * Dimension
        * Edit dimension
        * Subsets ^
          * Subset - edit
    * Chores ^
      * Chore - edit
    * Control Objects - filter everything } here
        * Control Cubes ^
        * Control Processes ^
        * Control Dimensions ^
    * Admin ^
      * Server Log
      * Threads
      * Configuration 
      * Run MDX
    * Git ^
 * Add Server -- should add the settings we need in the package json and register another top node
	
Required server parameters:
* username / password
* hostname
* integrated security mode
* (optional) folder to store model objects

**Comment by bgregs514** [comment]
This is definitely a good discussion topic 😃

I'm in full agreement with the Server, Cubes, Processes, Dimensions, Chores, and Control Objects parents.  What do you think about moving the Admin/Git parents as menu items in something like [this](https://code.visualstudio.com/api/extension-guides/tree-view#view-actions) (specifically the the view/title group: navigation menu).  Open to discussion on this one, but I was thinking the Git/Admin components were more of an "action", which may lend themselves more to some kind of sub-menu.

The contextValue in the TM1TreeItem class could be extended to have more additional logic baked into it so we could have different icons assigned depending on the type of object/actions that can be taken on it.  We could then have the "when" clause in the package.json toggle between different icons and states (these icons are actually attached to the "commands" tag, which is a bit odd).  Hopefully what I typed isn't too confusing, but an example would be extending the existing floppy disk save icon that is currently dynamically set on the TreeItems depending on if they are opened or not.

If we were to include the view/title and view/item/context menus into the tree, maybe it could look something like:

- Server ^
  - Cubes ^
    - Cube (inline "edit" pencil icon/possible structure icon?)
      - Views ^
        - View (inline "edit" pencil icon)
  - Processes ^
    - Process ^ (inline "run" icon (this time on the left before the object name like [this](https://github.com/microsoft/vscode/issues/103120#issuecomment-695355047) and inline "edit" pencil icon on right hand side)
  - Dimensions ^
    - Dimension ^ (inline "edit" pencil icon)
      - Subsets ^
        - Subset (inline "edit" pencil icon)
  - Chores ^
    - Chore (inline "edit" pencil icon)
  - Control Objects - filter everything } here
    - Control Cubes ^
    - Control Processes ^
    - Control Dimensions ^

This is just an idea, but basically the only difference between your proposed one would be to move the "actions" (i.e. run, edit, etc.) into different types of menus/buttons that sit either on the element itself, or in some kind of extension sub menu.

In regards to the parameters, that looks great!  The only thing I would possibly add is to allow for multiple hostnames.  The easiest solution for that would be to just serialize on ";" or something similar, but I wonder if there is a way to add a cleaner dynamic text box that allows for new entries to be added?  I couldn't really find anything [here](https://code.visualstudio.com/api/references/contribution-points#contributes.configuration), but I'll continue to look at other possibilities for this one.

As a future enhancement, we could look at adding "[welcome content](https://code.visualstudio.com/api/extension-guides/tree-view#welcome-content)" that directly links to the hostname settings if no hostname is specified on startup.  That's more of a nice-to-have, but could look kind of cool in the future.

What do you think?

Great job on that initial mockup!

**Reply by ykud** [reply]
Hi @bgregs514 , 

I agree your approach of adding buttons in general, I'd be keen to keep some of the objects like Rule icon for the sake of matching Architect / PAW editors to avoid people having to think that they need to press a pencil on a cube in this extension to get to a rule. The reason  I think Admin should be under server is for discoverability & context, but we can figure out the placement as we get closer to it. 

This layout in general also gives us a bit of a 'roadmap' and a clean separation of functionality, so once the common structure is in, somebody can work on presenting 'chores' independently to, for example, rules, so we can keep code changes concise. 

re multiple hostnames -- I don't fully understand why you'd want this. As far as I know TM1 server is on a unique host (can be on multiple admin servers, but only 1 host) and even when containerised version rolls in I'm sure it'll still be the host of container load balancer + port. 
So when I say Server ^, I mean a number of Server entries, like this:
* Planning Sample Dev
   * Cubes
   * Processes
   * ....
* Planning Sample Prod 
  * Cubes
  * Processes
  * ...
and therefore instead of multiple hostnames in one parameter, you end up with multiple configuration entries (one for each server) each having a hostname.

I was thinking of using quick input to collect parameters from user, something like this: 
https://github.com/microsoft/vscode-extension-samples/tree/main/quickinput-sample 
And a definite yes to a Welcome page saying 'add your server' :)

**Reply by bgregs514** [reply]
Perfect!  I'm with you - I misunderstood at first and meant to imply that we would have the different instances rolling up per admin host (basically exactly as you displayed in your example above).  I also confused "hostname" with "port number" in this case, my apologies 😄.

Also, given that we can query the HTTPPort number of individual instances with https://[admin_host]:[admin_port]/api/v1/Servers, we don't even need the individual instance HTTPPort numbers anyway.  That logic should be simple enough I think - the only thing I could potentially see is including the ability to "hide" certain instances that users may not be interested in.  They don't take up much space when consolidated, but having multiple instances can sometimes result in digging through the wrong one - maybe it's just me, but I do that _all the time_ in Architect.  That's a minor thing though, and definitely something we can address down the road (if it's even useful).  Another thought comes to mind, and that's "favoriting" specific instances so we can do stuff like "filter favorites" and only get the instances we care about.

I'm in agreement with having rules as a separate item under cubes; that approach probably aligns more with expected UI for users coming from Architect.  Hell, while we're at it, is there any benefit to adding a "favoriting" feature to a rule too?

In terms of the Admin parent, I see where you're coming from.  I do like the idea of having it front and center in the main TreeView so users can easily find it.  I'm fine with keeping the Admin node in the main TreeView at least as a starting point - it will be much easier to debug too since it won't be buried in a menu.  Once the logic is built around it, moving it around should be easy enough (in theory).

And finally, that quickinput dialog box is awesome!!  I didn't even know something like that existed.  So if I understand correctly, it would almost be like a guided wizard that would automatically set the configuration parameters in the user's settings.json file?  If that's the case, then I'm in 100% agreement that we should use that for setting up configs.

**Reply by ykud** [reply]
Hi @bgregs514 ,

I'm in between different meetings at the moment, will be a bit till I manage to put this properly, but I think the package json should include something like this:
```
"vscode-tm1.connectionList":{
					"type": "array",
                    "markdownDescription": "Server connections list",
                    "default": [],
					"items":{
						"type":"object",
						"properties":{
							"name": {
                                "type": "string",
                                "description": "Connection name"
                            },
							"hostname": {
								"type": "string",
								"default": "",
								"description": "TM1 host name"
							},	
							"httpportnumber": {
                                "type": "number",
                                "description": "HTTPPortNumber found in tm1s.cfg"
                            },
							"username": {
                                "type": "string",
                                "description": "Database username"
                            },
							"password": {
                                "type": [
                                    "string",
                                    "null"
                                ],
                                "description": "Connection password"
                            },
                            "askForPassword": {
                                "type": "boolean",
                                "description": "Ask for password instead of set it in your settings",
                                "default": false
                            },
							"integratedSecurityMode": {
								"type": "number",
								"default": 5,
								"enum": [1,2,3,4,5],
								"enumDescriptions": [
									"Basic: TM1 user name and password",
									"Basic + Integrated Login",
									"Integrated Login",
									"CAM without TM1 groups",
									"CAM with TM1 Groups"
								],
								"description": "TM1 Integrated Security Mode (1,2,3,4,5)"
							},
							"camNamespace": {
								"type": "string",
								"default": "",
								"description": "Namespace for CAM authentication"
							}
						}
					}
```

and we'd use the normal settings editor or something similar to provide it. 

Something like this substitutes the welcome page:
```
	"viewsWelcome":[{
			"view": "cubList",
			"contents": "No TM1 Server connections found.\n[Add new connection](command:vscode-tm1.openAddConnection)"
		}],
```

**Comment by TM1Fanboy** [comment]
Another thought could be to not make everything starting with a '}' a control object. We could show the ones which are not a real control objects in the list of 'normal' objects. In the end they are not control objects. They only exists because Architect hides those objects (and it is harder to delete them).

**Reply by ykud** [reply]
Simplicity of treating everything with } as a control object trumps the complexity of maintaining a list of rules of what is a 'real' control object in my book as you'd have to update these rule with every new type of control object added on the server.

As a middle ground we could group the known control objects into something like:
* Control objects
   * Security
      * Element
      * Cell
      * ...
   * Attributes
   * Stats
   * Other -- anything we didn't classify explicitly

**Reply by TM1Fanboy** [reply]
On the one hand I agree for simplicity but otherwise we would continue a practice which is confusing for newbies. The 'real' control objects get recreated on server startup, etc. the 'fake' ones are just normal cubes which are hidden in architect. 
The list shouldn't be to hard to maintain. New control cubes do not come that often and it should be an easy fix to add them to the list. 
It is a shame that there is no real flag in the rest api to distinguish those two types of }-cubes.

**Reply by rclapp** [reply]
There isn't a flag but there is an endpoint for this. Instead of cubes request ModelCubes (caps?). That should make this simple ☺

**Reply by bgregs514** [reply]
For an initial v1 release, I like the approach @rclapp posted to get a basic grouping of control cubes.  We can setup a simple json file behind the scenes to define categories to further group known cubes, but from an api standpoint I like the idea of calling "ModelCubes" instead of filtering on '}'.

Thanks for sharing, I wasn't aware that was even a possibility!

**Comment by TM1Fanboy** [comment]
One more...
I am missing hierarchies here. Should they be included (I think the answer is YES). 
What kind of views would you prefer? MDX or native views? Could make it a hard cut and only enable MDX views because they are the future anyway. On the other hand most people still rely on native views. 
Same goes for subsets by the way.

**Reply by tomcookactionkpi** [reply]
As someone who has to support and develop models that heavily leverage native views and subsets, my vote would be not to exclude native views and subsets.

Also a YES on the hierarchies.

**Comment by ykud** [comment]
Thought about it and I'd suggest that first phase would be only the code-related objects like Cube Rules / Processes / Chores, data driven things like views / dims / subsets might be too much to ask in the beginning as it opens the whole avenue of creating different viewers / editors and what not for them. 

A v1 release with:
* connecting to multiple servers with different modes of auth
* code editing for rules / tis / chores
* syntax highlighting
* basic IntelliSense for built-in functions (showing syntax and parameters)
would be a great accomplishment that is complimentary to what's out there already and we can see how it's extended in future.
v2 would be to hook it to Git :)

**Reply by TM1Fanboy** [reply]
Agree. Would go one step further and for v1 only rules and ti-code. That way we can show the potential of the tool and avoid the trap of having to add data sources and variables. Because:
- It is hard to add text files and derive the variables. 
- How should we know the existing ODBC connections?
- with views and subsets we get the same problem again. If I cannot create a dummy view for my data source, what should I add here. 
- We should ignore all other sources. 

While I am writing that, I had an opposing idea. If we have the whole code locally on file, we could create a list of existing data sources from the other processes and make that available as basis for data sources for new processes. Does that makes sense?

**Reply by ykud** [reply]
> * How should we know the existing ODBC connections
> While I am writing that, I had an opposing idea. If we have the whole code locally on file, we could create a list of existing data sources from the other processes and make that available as basis for data sources for new processes. Does that makes sense?

Just in case:
https://www.ibm.com/support/knowledgecenter/SSD29G_2.0.0/com.ibm.swg.ba.cognos.tm1_rest_api.2.0.0.doc/r_restapi_v1_csdl_alltypes.html#restapi_v1_csdl_entitytype_SQLDataSource

**Reply by TM1Fanboy** [reply]
> > * How should we know the existing ODBC connections
> 
> > While I am writing that, I had an opposing idea. If we have the whole code locally on file, we could create a list of existing data sources from the other processes and make that available as basis for data sources for new processes. Does that makes sense?
> 
> 
> 
> Just in case:
> 
> https://www.ibm.com/support/knowledgecenter/SSD29G_2.0.0/com.ibm.swg.ba.cognos.tm1_rest_api.2.0.0.doc/r_restapi_v1_csdl_alltypes.html#restapi_v1_csdl_entitytype_SQLDataSource
> 
> 

Okay. I admit I didn't know that. :-D 
But my other two points are maybe valid. 🤔

**Reply by bgregs514** [reply]
Great points and proposals from both!  I'm in full agreement with the v1 proposal, but I'm also a little torn when it comes to chores.  I would love to commit to rolling out chores as part of a v1 release, but I'm also a little cautious given that I just don't have much experience with working with chores and the rest API.  I can't imagine it's anything overly complex, but at the same time I feel that chore editing may lend itself more towards a wizard/UI type interface, and that is an area I'm not particularly great at gauging work effort.

If I had to give an answer, I'd lean more towards exclude chores from the initial v1 release notes, but begin to test it out behind the scenes to see if it's feasible.  That being said, if anyone is more comfortable with it and thinks it can be done along side rules/TIs, then by all means let's add it to the list!

Just my thoughts!

Side note: I also didn't know you could query the data source of a chore with the Rest API - very cool! 😉

**Reply by TM1Fanboy** [reply]
I am with you. Chores is basically UI only because you need a date picker. No one wants to write down the date and time for the chore manually. Besides that it is just listing processes and adding parameters for those. From my point of view there is nothing to code and all ui.

**Reply by ykud** [reply]
> I am with you. Chores is basically UI only because you need a date picker. No one wants to write down the date and time for the chore manually. Besides that it is just listing processes and adding parameters for those. From my point of view there is nothing to code and all ui. 

I was thinking about just a read only json chore file representation cause:
* I don't like disabling chores to see their schedule in Architect
* being able to search them in when looking for, say, a process name

In any case, nothing that urgent, let's see how we go with rules/tis )

**Reply by TM1Fanboy** [reply]
> I was thinking about just a read only json chore file representation cause:
> 
> * I don't like disabling chores to see their schedule in Architect

Very good point that is true for views and subsets as well. Maybe before we go into editing all those objects we first only show them Read-Only.
